### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by raghuhit**
Plugin for uploading app to app automate via fastlane.

### CODE REVIEW

File: lib/fastlane/plugin/browserstack/helper/browserstack_helper.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      # Params :
+      # +browserstack_username+:: BrowserStack's username.
+      # +browserstack_access_key+:: BrowserStack's access key.
+      # +file_path+:: Path to the file to be uploaded.
+      # +api_endpoint+:: BrowserStack's api endpoint, either AppAutomate or AppLive url.
+      def self.upload_file(browserstack_username, browserstack_access_key, file_path, api_endpoint)
+        url = "https://" + browserstack_username + ":" + browserstack_access_key + "@" + api_endpoint
+
+        begin
+          response = RestClient.post(url, file: File.new(file_path, 'rb'))
```
Comment (raj454raj) [inline lib/fastlane/plugin/browserstack/helper/browserstack_helper.rb]: Use something like [this](https://www.krautcomputing.com/blog/2015/06/21/how-to-use-basic-authentication-with-the-ruby-rest-client-gem/) for auth

File: lib/fastlane/plugin/browserstack/helper/browserstack_helper.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      end
+
+      # Uploads file to BrowserStack
+      # Params :
+      # +browserstack_username+:: BrowserStack's username.
+      # +browserstack_access_key+:: BrowserStack's access key.
+      # +file_path+:: Path to the file to be uploaded.
+      # +api_endpoint+:: BrowserStack's api endpoint, either AppAutomate or AppLive url.
+      def self.upload_file(browserstack_username, browserstack_access_key, file_path, api_endpoint)
+        url = "https://" + browserstack_username + ":" + browserstack_access_key + "@" + api_endpoint
```
Comment (raj454raj) [inline lib/fastlane/plugin/browserstack/helper/browserstack_helper.rb]: move this https to where the api_endpoint is initialized

**Comment by raghuhit** [comment]
@raj454raj I have made the changes which you asked for, please review.

File: lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+require_relative '../helper/browserstack_helper'
+require 'json'
+
+module Fastlane
+  module Actions
+    module SharedValues
+      BROWSERSTACK_APP_ID ||= :BROWSERSTACK_APP_ID
+    end
+    class UploadToBrowserstackAppAutomateAction < Action
+      SUPPORTED_FILE_EXTENSIONS = ["apk", "ipa", "zip"]
```
Comment (kunal361) [inline lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb]: Supported are just apk and ipa for the below endpoint.

If this plugin has to support espresso and xcui (and earlgrey in the future) as well, then we need to support zip. In that case, the api endpoints will change based on the framework.

File: lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        Actions.lane_context[SharedValues::BROWSERSTACK_APP_ID] = browserstack_app_id
+      end
+
+      # Validate file_path.
+      def self.validate_file_path(file_path)
+        UI.user_error!("No file found at '#{file_path}'.") unless File.exist?(file_path)
+
+        # Validate file extension.
+        file_path_parts = file_path.split(".")
+        unless file_path_parts.length == 2 && SUPPORTED_FILE_EXTENSIONS.include?(file_path_parts[1])
```
Comment (kunal361) [inline lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb]: There is no restriction on the number of `.`s that can be there in a filename. Use the last part to verify file extension.

File: fastlane/Fastfile [hunk 5]
```diff
@@ -0,0 +1,7 @@
+lane :test do
+  upload_to_browserstack_app_automate(
+    browserstack_username: ENV["BROWSERSTACK_USERNAME"],
+    browserstack_access_key: ENV["BROWSERSTACK_ACCESS_KEY"],
+    file_path: "path_to_apk_or_ipa_or_zip_file"
```
Comment (kunal361) [inline fastlane/Fastfile]: there should be a way to programmatically determine the file path.

File: README.md [hunk 6]
```diff
@@ -1 +1,49 @@
...
+
+This project is a [_fastlane_](https://github.com/fastlane/fastlane) plugin. To get started with `fastlane-plugin-browserstack`, add it to your project by running:
+
+```bash
+fastlane add_plugin browserstack
+```
+
+## About browserstack
+
+Uploads IPA, APK and ZIP files to BrowserStack for manual and automation testing.
```
Comment (kunal361) [inline README.md]: manual testing isnt supported in this version right?

File: lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb (hunk 3 above)
Comment (raghuhit) [reply lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb]: What is the general endpoint for uploading all kinds of app for testing?

Comment (kunal361) [reply lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb]: there isn't 1 endpoint.
ipa/apk apps: https://api-cloud.browserstack.com/app-automate/upload
apk (espresso test suite): https://api-cloud.browserstack.com/app-automate/espresso/test-suite
zip (xcui test suite): https://api-cloud.browserstack.com/app-automate/xcuitest/test-suite
zip (earlgrey test suite): https://api-cloud.browserstack.com/app-automate/earlgrey/app-dir

Comment (raghuhit) [reply lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb]: Currently, we are going to support apk and ipa file upload only, hence using first endpoint : https://api-cloud.browserstack.com/app-automate/upload .

File: lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb (hunk 4 above)
Comment (raghuhit) [reply lib/fastlane/plugin/browserstack/actions/upload_to_browserstack_app_automate_action.rb]: Done

**Pull request merged**